src/libraries/System.Threading.Channels/src/System/Threading/Channels/ChannelUtilities.cs

```diff
@@ -66,60 +66,181 @@ internal static ValueTask<T> GetInvalidCompletionValueTask<T>(Exception error)
             return new ValueTask<T>(t);
         }
 
-        internal static void QueueWaiter(ref AsyncOperation<bool>? tail, AsyncOperation<bool> waiter)
+        /// <summary>Dequeues an operation from the circular doubly-linked list referenced by <paramref name="head"/>.</summary>
+        /// <param name="head">The head of the list.</param>
+        /// <param name="op">The dequeued operation.</param>
+        /// <returns>true if an operation could be dequeued; otherwise, false.</returns>
+        internal static bool TryDequeue<TAsyncOp>(ref TAsyncOp? head, [NotNullWhen(true)] out TAsyncOp? op)
+            where TAsyncOp : AsyncOperation<TAsyncOp>
         {
-            AsyncOperation<bool>? c = tail;
-            if (c == null)
+            op = head;
+
+            if (head is null)
+            {
+                return false;
+            }
+
+            Debug.Assert(head.Previous is not null);
+            Debug.Assert(head.Next is not null);
+
+            if (head.Next == head)
+            {
+                head = null;
+            }
+            else
+            {
+                TAsyncOp last = head.Previous;
+
+                head = head.Next;
+                head.Previous = last;
+                last.Next = head;
+            }
+
+            Debug.Assert(op is not null);
+            op.Next = op.Previous = null;
+            return true;
+        }
+
+        /// <summary>Enqueues an operation onto the circular doubly-linked list referenced by <paramref name="head"/>.</summary>
+        /// <param name="head">The head of the list.</param>
+        /// <param name="op">The operation to enqueue.</param>
+        internal static void Enqueue<TAsyncOp>(ref TAsyncOp? head, TAsyncOp op)
+            where TAsyncOp : AsyncOperation<TAsyncOp>
+        {
+            Debug.Assert(op.Next is null && op.Previous is null);
+
+            if (head is null)
+            {
+                head = op.Next = op.Previous = op;
+            }
+            else
+            {
+                TAsyncOp last = head.Previous!;
+                Debug.Assert(last is not null);
+
+                op.Next = head;
+                op.Previous = last;
+                last.Next = op;
+                head.Previous = op;
+            }
+        }
+
+        /// <summary>Removes the specified operation from the circular doubly-linked list referenced by <paramref name="head"/>.</summary>
+        /// <param name="head">The head of the list.</param>
+        /// <param name="op">The operation to remove.</param>
+        internal static void Remove<TAsyncOp>(ref TAsyncOp? head, TAsyncOp op)
+            where TAsyncOp : AsyncOperation<TAsyncOp>
+        {
+            Debug.Assert(op is not null);
+            Debug.Assert(op.Next is null == op.Previous is null);
```

Comment:
I think this can have the same/similar race. This line isn't atomic:

```cs
current.Next = current.Previous = null;
```

ChannelUtilities, line 198